federatedscope/core/data/base_translator.py

```diff
@@ -0,0 +1,123 @@
+import logging
+import numpy as np
+
+from federatedscope.core.auxiliaries.splitter_builder import get_splitter
+from federatedscope.core.data import ClientData, StandaloneDataDict
+
+logger = logging.getLogger(__name__)
+
+
+class BaseDataTranslator:
+    """
+    Perform process:
+        Dataset -> ML split -> FL split -> Data (passed to FedRunner)
+
+    """
+    def __init__(self, global_cfg, client_cfgs=None):
+        """
+        Convert data to `StandaloneDataDict`.
+
+        Args:
+            global_cfg: global CfgNode
+            client_cfgs: client cfg `Dict`
+        """
+        self.global_cfg = global_cfg.clone()
+        self.client_cfgs = client_cfgs
+        self.splitter = get_splitter(global_cfg)
+
+    def __call__(self, dataset):
+        """
+
+        Args:
+            dataset: `torch.utils.data.Dataset`, `List` of (feature, label)
+                or split dataset tuple of (train, val, test) or Tuple of
+                split dataset with [train, val, test]
+        """
+        datadict = self.split(dataset)
+        datadict = StandaloneDataDict(datadict, self.global_cfg)
+
+        return datadict
+
+    def split(self, dataset):
+        """
+        Perform ML split and FL split.
+
+        Returns:
+            dict of `ClientData` with client_idx as key.
+
+        """
+        train, val, test = self.split_train_val_test(dataset)
+        datadict = self.split_to_client(train, val, test)
+        return datadict
+
+    def split_train_val_test(self, dataset):
+        """
+        Split dataset to train, val, test if not provided.
+
+        Returns:
+            split_data (List): List of split dataset, [train, val, test]
+
+        """
+        splits = self.global_cfg.data.splits
+        if isinstance(dataset, tuple):
+            # No need to split train/val/test for tuple dataset.
+            error_msg = 'If dataset is tuple, it must contains ' \
+                        'train, valid and test split.'
+            assert len(dataset) == len(['train', 'val', 'test']), error_msg
+            return [dataset[0], dataset[1], dataset[2]]
+
+        index = np.random.permutation(np.arange(len(dataset)))
```

Comment:
Could you teach me how to use the default splits (e.g., OGB)? I have not found the related code snippets.

Reply:
```python
trans = BaseDataTranslator(cfg)
datadict = trans((train_data, val_data, test_data))
```